client/asset/dcr/dcr.go

```diff
@@ -1231,141 +1256,255 @@ func (dcr *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, time.Time,
+			// Redemption search for this set of contracts is compromised. Notify
+			// the redemption finder(s) of this fatal error and cancel redemption
+			// search for these contracts. The redemption finder(s) may re-call
+			// dcr.FindRedemption to restart find redemption attempts for any of
+			// these contracts.
+			err = fmt.Errorf("error fetching verbose block %s: %v", nextBlockHash, err)
+			dcr.fatalFindRedemptionsError(err, contractOutpoints)
+			return
 		}
-		// If this is the swap contract's block, grab the contract hash.
-		if *blockHash == contractBlock {
-			for i := range block.RawTx {
-				if block.RawTx[i].Txid == txid {
-					vouts := block.RawTx[i].Vout
-					if len(vouts) < int(vout)+1 {
-						return nil, fmt.Errorf("vout %d not found in tx %s", vout, txid)
-					}
-					contractHash, err = dexdcr.ExtractContractHash(vouts[vout].ScriptPubKey.Hex)
-					if err != nil {
-						return nil, err
-					}
-				}
+		scanPoint := fmt.Sprintf("block %d", blk.Height)
+		lastScannedBlockHeight = blk.Height
+		blkTxs := append(blk.RawTx, blk.RawSTx...)
+		for t := range blkTxs {
+			tx := &blkTxs[t]
+			redemptionsFound += dcr.findRedemptionsInTx(scanPoint, tx, contractOutpoints)
+			if redemptionsFound == contractsCount {
+				break rangeBlocks
 			}
 		}
-		if len(contractHash) == 0 {
-			return nil, fmt.Errorf("no secret hash found at %s:%d", txid, vout)
+		if blk.NextHash == "" {
+			nextBlockHash = nil
+		} else {
+			nextBlockHash, err = chainhash.NewHashFromStr(blk.NextHash)
+			if err != nil {
+				err = fmt.Errorf("hash decode error %s: %v", blk.NextHash, err)
+				dcr.fatalFindRedemptionsError(err, contractOutpoints)
+				return
+			}
 		}
-		// Look at the previous outpoint of each input of each transaction in the
-		// block.
-		for i := range block.RawTx {
-			for j := range block.RawTx[i].Vin {
-				if block.RawTx[i].Vin[j].Txid == txid && block.RawTx[i].Vin[j].Vout == vout {
-					// Found it. Extract the key.
-					vin := block.RawTx[i].Vin[j]
-					sigScript, err := hex.DecodeString(vin.ScriptSig.Hex)
-					if err != nil {
-						return nil, fmt.Errorf("error decoding scriptSig '%s': %v", vin.ScriptSig.Hex, err)
-					}
-					key, err := dexdcr.FindKeyPush(sigScript, contractHash, chainParams)
-					if err != nil {
-						return nil, fmt.Errorf("error extracting key from '%s': %v", vin.ScriptSig.Hex, err)
-					}
-					return key, nil
+	}
+
+	dcr.findRedemptionMtx.RLock()
+	contractsInQueue := len(dcr.findRedemptionQueue)
+	dcr.findRedemptionMtx.RUnlock()
+
+	if redemptionsFound > 0 {
+		dcr.log.Debugf("found redemptions for %d contracts in blocks %d - %d, %d contracts left in queue",
+			redemptionsFound, startBlock.height, lastScannedBlockHeight, contractsInQueue)
+	} else {
+		dcr.log.Debugf("redemptions not found for %d contracts in blocks %d - %d, %d contracts left in queue",
+			contractsCount, startBlock.height, lastScannedBlockHeight, contractsInQueue)
+	}
```

Comment:
I would combine these as they are almost the same. `n redemptions out of n contracts found in blocks...`